db/commitlog/commitlog.cc

```diff
@@ -1713,10 +1716,9 @@ future<> db::commitlog::segment_manager::shutdown() {
                 return std::move(_reserve_replenisher).then_wrapped([this] (auto f) {
                     // Could be cleaner with proper seastar support
                     if (f.failed()) {
-                        _shutdown_promise->set_exception(f.get_exception());
-                    } else {
-                        _shutdown_promise->set_value();
+                        clogger.error("Reserve replenisher failed during shutdown: {}. Continuing with shutdown anyway...", f.get_exception());
```

Comment:
I see that replenish_reserve() already handles errors.

Reply:
btw, maybe badly, it will ignore a gate closed exception.

Reply:
you're right.  yet there is existing error handling code here that propagates the impossible to reach error to _shutdown_promise.

I'll remove this hunk.

Reply:
But let's fix replenish not to ignore the gate closed. Maybe that's the root cause.

Reply:
I don't think that's the root cause since
https://github.com/scylladb/scylla/blob/4c351ff26009ece329231560cbcb0b54a468da0a/db/commitlog/commitlog.cc#L1179
will throw in the then clause and the the error will propagate as an exceptional future here (that's a subtle difference between `with_gate` and `try_with_gate`, so it's not ignoring a possible gate_closed exception.

If it did, and with `_shutdown` not set to true in case `sync_all_segments` fails, it could stay in its `do_until` loop, but then we should see a warning coming from https://github.com/scylladb/scylla/blob/4c351ff26009ece329231560cbcb0b54a468da0a/db/commitlog/commitlog.cc#L1193
but we see no error or warning, according to https://github.com/scylladb/scylla/issues/8577#issue-874223808

Reply:
Ah, yes. Those gotchas are terrible. Let's rewrite it as a coroutine (after we fix it).